ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,536 @@
+   by the Free Software Foundation; version 2.1 only. with the special
+   exception on linking described in file LICENSE.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+*)
+
+let ( // ) = Filename.concat
+
+let runtime_data = "/var/lib" // "xapi-guard"
+
+let ( let* ) = Lwt.bind
+
+let ( let@ ) f x = f x
+
+let with_lock = Lwt_mutex.with_lock
+
+type t = Uuidm.t * Mtime.t * Types.Tpm.key
+
+let cache_of service = runtime_data // Types.Service.to_string service
+
+let files_in dir ~otherwise =
+  Lwt.catch
+    (fun () ->
+      let* listing = Lwt_unix.files_of_directory dir |> Lwt_stream.to_list in
+      List.filter_map
+        (function "." | ".." -> None | name -> Some (dir // name))
+        listing
+      |> Lwt.return
+    )
+    otherwise
+
+let unlink_safe file =
+  let __FUN = __FUNCTION__ in
+  Lwt.catch
+    (fun () -> Lwt_unix.unlink file)
+    (function
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+          Lwt.pause ()
+      | e ->
+          Logs_lwt.info (fun m ->
+              m "%s: error %s when deleting %s, ignoring" __FUN
+                (Printexc.to_string e) file
+          )
+      )
+
+type valid_file = t * string
+
+type file =
+  | Latest of valid_file
+  | Outdated of valid_file
+  | Temporary of string
+  | Invalid of string
+
+let path_of_key root (uuid, timestamp, key) =
+  root
+  // Uuidm.to_string uuid
+  // Types.Tpm.(serialize_key key |> string_of_int)
+  // Mtime.(to_uint64_ns timestamp |> Int64.to_string)
+
+let key_of_path path =
+  let ( let* ) = Option.bind in
+  let key_dir = Filename.(dirname path) in
+  let* uuid = Filename.(basename (dirname key_dir)) |> Uuidm.of_string in
+  let* key =
+    Filename.basename key_dir
+    |> int_of_string_opt
+    |> Option.map Types.Tpm.deserialize_key
+  in
+  let* timestamp =
+    Filename.basename path
+    |> Int64.of_string_opt
+    |> Option.map Mtime.of_uint64_ns
+  in
+  Some ((uuid, timestamp, key), path)
+
+let path_is_temp path =
+  let pathlen = String.length path in
+  String.ends_with ~suffix:".pre" path
+  && key_of_path (String.sub path 0 (pathlen - 4)) |> Option.is_some
+
+let temp_of_path path = path ^ ".pre"
+
+let get_all_contents root =
+  let classify contents =
+    let rec loop (acc, found) = function
+      | [] ->
+          List.rev acc
+      | latest :: others -> (
+        match key_of_path latest with
+        | None ->
+            let file =
+              if path_is_temp latest then
+                Temporary latest
+              else
+                Invalid latest
+            in
+            loop (file :: acc, found) others
+        | Some valid_file ->
+            let file =
+              if found then Outdated valid_file else Latest valid_file
```

Comment:
Who sets `found`? AFAICT it is always `false`

Reply:
It should be set in the following line, because a valid file was found, fixed